src/main/java/org/prebid/server/auction/BidResponseCreator.java

```diff
@@ -512,6 +522,57 @@ private Bid toBid(BidderBid bidderBid, String bidder, ExtRequestTargeting target
         return bid;
     }
 
+    private static void addNativeMarkup(Bid bid, List<Imp> imps) {
+        final Response nativeMarkup;
+        try {
+            nativeMarkup = Json.decodeValue(bid.getAdm(), Response.class);
+        } catch (DecodeException e) {
+            throw new PreBidException(e.getMessage());
+        }
+
+        final List<Asset> responseAssets = nativeMarkup.getAssets();
+        if (CollectionUtils.isNotEmpty(responseAssets)) {
+            final Native nativeImp = imps.stream()
+                    .filter(imp -> imp.getId().equals(bid.getImpid()) && imp.getXNative() != null)
+                    .findFirst()
+                    .map(Imp::getXNative)
+                    .orElseThrow(() -> new PreBidException("Could not find native imp"));
+
+            final Request nativeRequest;
+            try {
+                nativeRequest = Json.mapper.readValue(nativeImp.getRequest(), Request.class);
+            } catch (JsonProcessingException e) {
+                throw new PreBidException(e.getMessage());
+            }
+
+            responseAssets.forEach(asset -> setAssetTypes(asset, nativeRequest.getAssets()));
+            bid.setAdm(Json.encode(nativeMarkup));
+        }
+    }
+
+    private static void setAssetTypes(Asset responseAsset, List<com.iab.openrtb.request.Asset> requestAssets) {
+        if (responseAsset.getImg() != null) {
+            final Integer type = getAssetById(responseAsset.getId(), requestAssets).getImg().getType();
```

Comment:
Does we guarantee that `getImg` always be not null. Bc you can return empty `Asset` from `getAssetById`

Reply:
Good point, thank you!